daemon/cluster/cluster.go

```diff
@@ -871,6 +872,34 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
 	return services, nil
 }
 
+// imageWithDigestString takes an image such as name or name:tag
+// and returns the image pinned to a digest, such as name@sha256:34234...
+func (c *Cluster) imageWithDigestString(ctx context.Context, image, encodedAuth string) (string, error) {
+	// TODO(nishanttotla): create AuthConfig from encodedAuth
+
+	ref, err := reference.ParseNamed(image)
+	if err != nil {
+		logrus.Warnf("unable to parse image name: %s", err.Error())
+		return "", err
+	}
+	// only pin if not a canonical reference (i.e. with digest)
+	if _, ok := ref.(reference.Canonical); !ok {
+		// add default tag if not present
+		if reference.IsNameOnly(ref) {
+			ref = reference.WithDefaultTag(ref)
+		}
```

Comment:
The `if` is not necessary - you can just do `ref = reference.WithDefaultTag(ref)` since `WithDefaultTag` is defined as follows:

```go
// WithDefaultTag adds a default tag to a reference if it only has a repo name.
func WithDefaultTag(ref Named) Named {
        if IsNameOnly(ref) {
                ref, _ = WithTag(ref, DefaultTag)
        }
        return ref
}
```